File: packages/xyd-openapi2opencli/src/schema.ts
import type { OpenAPIV3 } from 'openapi-types';

export type Schema = OpenAPIV3.SchemaObject;

const SCALAR_TYPES = new Set(['string', 'number', 'integer', 'boolean']);

/** The scalar type of a schema, or null if it isn't a scalar. */
export function scalarType(schema: Schema | undefined): string | null {
  if (!schema) return null;
  const t = typeof schema.type === 'string' ? schema.type : undefined;
  if (t && SCALAR_TYPES.has(t)) return t;
  return null;
}

export function isScalar(schema: Schema | undefined): boolean {
  return scalarType(schema) !== null;
}

export function isBoolean(schema: Schema | undefined): boolean {
  return scalarType(schema) === 'boolean';
}

export function isArray(schema: Schema | undefined): boolean {
  return !!schema && schema.type === 'array';
}

export function arrayItems(schema: Schema | undefined): Schema | undefined {
  if (!schema || schema.type !== 'array') return undefined;
  return schema.items as Schema | undefined;
}

/** True when the schema is an object (explicit type or has properties). */
export function isObjectSchema(schema: Schema | undefined): boolean {
  if (!schema) return false;
  if (schema.type === 'object') return true;
  if (schema.properties && Object.keys(schema.properties).length > 0) return true;
  return false;
}

/** Enum values rendered as strings (undefined when not an enum). */
export function getEnum(schema: Schema | undefined): string[] | undefined {
  if (!schema || !Array.isArray(schema.enum) || schema.enum.length === 0) return undefined;
  return schema.enum.map((v) => String(v));
}

export function getDefault(schema: Schema | undefined): unknown {
  if (!schema) return undefined;
  return (schema as { default?: unknown }).default;
}

export function getDescription(schema: Schema | undefined): string | undefined {
  return schema?.description;
}

/**
 * Merge an `allOf` chain into a single object schema (properties + required).
 * Non-allOf schemas are returned unchanged.
 */
export function mergeAllOf(schema: Schema | undefined, seen: WeakSet<object> = new WeakSet()): Schema | undefined {
  if (!schema || !Array.isArray(schema.allOf) || schema.allOf.length === 0) return schema;
  // Guard against allOf cycles in dereferenced (possibly circular) schemas.
  if (seen.has(schema)) return schema;
  seen.add(schema);

  const merged: Schema = { type: 'object', properties: {}, required: [] };
  const required = new Set<string>(Array.isArray(schema.required) ? schema.required : []);

  for (const sub of schema.allOf as Schema[]) {
    const resolved = mergeAllOf(sub, seen) ?? sub;
    if (resolved.properties) {
      Object.assign(merged.properties as Record<string, unknown>, resolved.properties);
    }
    if (Array.isArray(resolved.required)) {
      for (const r of resolved.required) required.add(r);
    }
  }
  // Properties declared directly alongside allOf
  if (schema.properties) {
    Object.assign(merged.properties as Record<string, unknown>, schema.properties);
  }
  merged.required = [...required];
  if (schema.description) merged.description = schema.description;
  return merged;
}

/**
 * Resolve a body schema to the object schema we can flatten, plus a flag telling
 * whether the original was a non-trivial composition (oneOf/anyOf) we only
 * partially captured.
 */
export function resolveObjectSchema(schema: Schema | undefined): { object?: Schema; complex: boolean } {
  if (!schema) return { complex: false };

  const merged = mergeAllOf(schema) ?? schema;
  if (isObjectSchema(merged)) return { object: merged, complex: false };

  // oneOf / anyOf at the body root: take the first object branch, mark complex.
  const branches = (merged.oneOf || merged.anyOf) as Schema[] | undefined;
  if (Array.isArray(branches)) {
    for (const branch of branches) {
      const b = mergeAllOf(branch) ?? branch;
      if (isObjectSchema(b)) return { object: b, complex: true };
    }
    return { complex: true };
  }

  return { complex: false };
}

/** True for a binary/file string property (multipart uploads). */
export function isBinary(schema: Schema | undefined): boolean {
  return !!schema && schema.type === 'string' && schema.format === 'binary';
}

File: crates/xyd_openapi2opencli/src/schema.rs
//! schema.rs helpers — port of the OpenCLI schema.ts. Operates on already
//! `$ref`-resolved schema values (`ctx.resolve()` is applied at the call
//! sites); `type` is read as a plain string (the openapi-types 3.0 shape the
//! JS assumes — a 3.1 array-type yields "not a scalar", matching JS).

use serde_json::{Map, Value};
use std::collections::HashSet;

use crate::jsrt::js_string;

const SCALAR_TYPES: [&str; 4] = ["string", "number", "integer", "boolean"];

fn type_str(schema: Option<&Value>) -> Option<&str> {
    // JS `typeof schema.type === 'string' ? schema.type : undefined`.
    schema?.get("type")?.as_str()
}

pub fn scalar_type(schema: Option<&Value>) -> Option<&'static str> {
    let t = type_str(schema)?;
    SCALAR_TYPES.iter().copied().find(|&s| s == t)
}

pub fn is_boolean(schema: Option<&Value>) -> bool {
    scalar_type(schema) == Some("boolean")
}

pub fn is_array(schema: Option<&Value>) -> bool {
    type_str(schema) == Some("array")
}

pub fn array_items(schema: Option<&Value>) -> Option<&Value> {
    let s = schema?;
    if s.get("type").and_then(|t| t.as_str()) != Some("array") {
        return None;
    }
    s.get("items")
}

/// Object by explicit `type: object` or by having non-empty properties.
pub fn is_object_schema(schema: Option<&Value>) -> bool {
    let Some(s) = schema else { return false };
    if s.get("type").and_then(|t| t.as_str()) == Some("object") {
        return true;
    }
    matches!(s.get("properties"), Some(Value::Object(p)) if !p.is_empty())
}

/// Enum values rendered as strings via JS `String(v)`; None when not an enum.
pub fn get_enum(schema: Option<&Value>) -> Option<Vec<String>> {
    let arr = schema?.get("enum")?.as_array()?;
    if arr.is_empty() {
        return None;
    }
    Some(arr.iter().map(js_string).collect())
}

pub fn get_default(schema: Option<&Value>) -> Option<&Value> {
    schema?.get("default")
}

pub fn is_binary(schema: Option<&Value>) -> bool {
    let Some(s) = schema else { return false };
    s.get("type").and_then(|t| t.as_str()) == Some("string")
        && s.get("format").and_then(|f| f.as_str()) == Some("binary")
}

/// `mergeAllOf` — merge an allOf chain into one object schema. Non-allOf
/// returns a clone. Address-based cycle guard (the JS WeakSet).
pub fn merge_all_of(schema: &Value, seen: &mut HashSet<usize>) -> Value {
    let all_of = match schema.get("allOf") {
        Some(Value::Array(arr)) if !arr.is_empty() => arr,
        _ => return schema.clone(),
    };
    let addr = schema as *const Value as usize;
    if seen.contains(&addr) {
        return schema.clone();
    }
    seen.insert(addr);

    let mut props: Map<String, Value> = Map::new();
    let mut required: Vec<String> = Vec::new();
    let mut required_seen: HashSet<String> = HashSet::new();

    for sub in all_of {
        let resolved = merge_all_of(sub, seen);
        if let Some(Value::Object(p)) = resolved.get("properties") {
            for (k, v) in p {
                props.insert(k.clone(), v.clone());
            }
        }
        if let Some(Value::Array(req)) = resolved.get("required") {
            for r in req {
                if let Some(rs) = r.as_str() {
                    if required_seen.insert(rs.to_string()) {
                        required.push(rs.to_string());
                    }
                }
            }
        }
    }
    if let Some(Value::Object(p)) = schema.get("properties") {
        for (k, v) in p {
            props.insert(k.clone(), v.clone());
        }
    }

    let mut merged = Map::new();
    merged.insert("type".into(), Value::String("object".into()));
    merged.insert("properties".into(), Value::Object(props));
    merged.insert(
        "required".into(),
        Value::Array(required.into_iter().map(Value::String).collect()),
    );
    if let Some(desc) = schema.get("description") {
        merged.insert("description".into(), desc.clone());
    }
    Value::Object(merged)
}

pub struct ResolvedBody {
    pub object: Option<Value>,
    #[allow(dead_code)]
    pub complex: bool,
}

/// `resolveObjectSchema` — merge allOf, take the object; else the first object
/// branch of oneOf/anyOf (marked complex).
pub fn resolve_object_schema(schema: Option<&Value>) -> ResolvedBody {
    let Some(schema) = schema else {
        return ResolvedBody {
            object: None,
            complex: false,
        };
    };
    let mut seen = HashSet::new();
    let merged = merge_all_of(schema, &mut seen);
    if is_object_schema(Some(&merged)) {
        return ResolvedBody {
            object: Some(merged),
            complex: false,
        };
    }
    let branches = merged
        .get("oneOf")
        .or_else(|| merged.get("anyOf"))
        .and_then(|b| b.as_array());
    if let Some(branches) = branches {
        for branch in branches {
            let mut s = HashSet::new();
            let b = merge_all_of(branch, &mut s);
            if is_object_schema(Some(&b)) {
                return ResolvedBody {
                    object: Some(b),
                    complex: true,
                };
            }
        }
        return ResolvedBody {
            object: None,
            complex: true,
        };
    }
    ResolvedBody {
        object: None,
        complex: false,
    }
}
